Stratis.Bitcoin.Features.Consensus/ConsensusLoop.cs

```diff
@@ -62,78 +164,188 @@ public void Initialize()
                         }
+
+                        this.Tip = rewinded;
+                        this.Puller.SetLocation(rewinded);
+                        this.logger.LogInformation("Reorg detected, rewinding from '{0}' to '{1}'.", lastTip, this.Tip);
+                        return;
                     }
                 }
 
-                this.AcceptBlock(new ContextInformation(result, this.Validator.ConsensusParams));
-            }
-            catch (ConsensusErrorException ex)
-            {
-                result.Error = ex.ConsensusError;
+                try
+                {
+                    this.ValidateBlock(new ContextInformation(item, this.Validator.ConsensusParams));
+                }
+                catch (ConsensusErrorException ex)
+                {
+                    item.Error = ex.ConsensusError;
+                }
+
+                if (item.Error != null)
+                {
+                    this.logger.LogError("Block rejected: {0}", item.Error.Message);
+
+                    // Pull again.
+                    this.Puller.SetLocation(this.Tip);
+
+                    if (item.Error == ConsensusErrors.BadWitnessNonceSize)
+                    {
+                        this.logger.LogInformation("You probably need witness information, activating witness requirement for peers.");
+                        this.connectionManager.AddDiscoveredNodesRequirement(NodeServices.NODE_WITNESS);
+                        this.Puller.RequestOptions(TransactionOptions.Witness);
+                        return;
+                    }
+
+                    // Set the chain back to ConsensusLoop.Tip.
+                    this.Chain.SetTip(this.Tip);
+                    this.logger.LogTrace("Chain reverted back to block '{0}' ", this.Tip);
+
+                    // Since ChainHeadersBehavior check PoW, MarkBlockInvalid can't be spammed.
+                    this.logger.LogError("Marking block as invalid.");
+                    this.chainState.MarkBlockInvalid(item.Block?.GetHash());
+                }
+                else
+                {
+                    this.logger.LogTrace("Block accepted '{0}' ", this.Tip);
+
+                    this.chainState.HighestValidatedPoW = this.Tip;
+
+                    // We really want to flush if we are at the top of the chain.
+                    // Otherwise, we just allow the flush to happen if it is needed.
+                    bool forceFlush = this.Chain.Tip.HashBlock == item.ChainedBlock?.HashBlock;
+                    this.FlushAsync(forceFlush).GetAwaiter().GetResult();
+
+                    if (this.Tip.ChainWork > this.Chain.Tip.ChainWork)
+                    {
+                        // This is a newly mined block.
+                        this.Chain.SetTip(this.Tip);
+                        this.Puller.SetLocation(this.Tip);
+
+                        this.logger.LogDebug("Block extends consensus tip to '{0}'", this.Tip);
+                    }
```

Comment:
this branch does not really prefer the blocks that we mined as we discussed it earlier, right?

however, if we really think about what will happen exactly, it does not sound right:
- first we receive next block from the peer
- block goes to this method, is being validated
- then we mine a new block, block goes to this method, waits for lock to enter
- peer block is finished and leaves the method
- mined block goes in critical sections
- item.Block is not null, so first part is skipped
- then we go to ValidateBlock where `context.BlockItem.Block.Header.HashPrevBlock != this.Tip.HashBlock` this condition will be true, so we should get `ConsensusErrors.InvalidPrevTip.Throw`
- we go to `Block rejected` branch in AcceptBlock

so we would actually treat a perfectly valid block as invalid block

in previous discussion i've suggested that such a block should instead replace the current tip and we should try to prolong that chain as it seems more beneficial to us than other strategy. maybe that is more complicated and not really necessary to do, so i'm not saying we need to rewrite it as such

but having it jump to Block rejected seems like a mistake. i don't think anything bad happen, that block will be marked as invalid, which just occupies some memory, but except for that no harm will be done. conceptually it feels wrong however - maybe a simple check on whether that InvalidPrevTip was received and handle it gracefully would feel better? i.e. just do nothing instead of going to Block rejected branch

Reply:
ok, so the above scenario is just weird, but INVERSE scenario is actually dangerous. this is serious bug and MUST be fixed

Reply:
OK, so to just explain that - in the first post here I've analyzed the situation when you first get block from peer and immediately after we mine a block. That's not optimal, but it is handled +- OK without much harm. But the inverse situation is very bad - if we mine a block and just after that we receive block from peer, it seems there is a disaster. That block from peer will be marked as invalid and it is very good chance that the real chain will be built on the top of it (and not our block). Then our node is out of the chain and can't go back.

Reply:
Ah yes this is a big but however the mark as invalid code is not used (yet).
But you are right we should probably verify the reason of failure if the reason is only less work then to not mark the block as invalid, but I guess this should be done in the banning PR.

Reply:
actually I think this is used already in CHB,
and even if it was not, this is a bug problem and i would say it must be fixed in this PR
the fix is rather trivial, so no need not to do it
that situation will be very often (esp. on testnet)

Reply:
Where is it used? CHB does not act on invalid blocks.
But ok as per your request I'll remove it.

Reply:
i can see it in headerspayload processing: 
`validated &= !this.SharedState.IsMarkedInvalid(tip.HashBlock);`

Reply:
OK yeah I see, the fix in this particular case will be to not mark as invalid if the reason is only work.
So if the error is InvalidPreviousBlock then will not mark the block as invalid, how ever when banning is implemented each error will have a severity then the severity will need to be used..

Reply:
yes, just handle InvalidPrevTip differently
it is not about work, it is about whether the block that arrived prolongs what we consider the best chain or not and as we shown it can be the case that valid block that does not prolong the best chain is processed here

Reply:
You normally prolong the chain by providing proof you made enough work to prolong it (or in this case stake)